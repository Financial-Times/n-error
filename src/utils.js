export const removeObjectKeys = obj => keys => {
	const output = obj;
	if (Array.isArray(keys)) {
		Object.keys(obj).forEach(key => {
			if (keys.includes(key)) {
				delete output[key];
			}
		});
		// if (obj instanceof Error) {
		// 	['message', 'stack', 'code'].forEach(key => {
		// 		if (keys.includes(key)) {
		// 			Object.assign(obj, { [key]: undefined });
		// 		}
		// 	});
		// }
		return output;
	} else if (typeof keys === 'string' && keys !== '') {
		const k = keys;
		Object.keys(obj).forEach(key => {
			if (key === k) {
				delete output[key];
			}
		});
		// if (obj instanceof Error && ['message', 'stack', 'code'].includes(k)) {
		// 	Object.assign(output, { [k]: undefined });
		// }
		return output;
	}
	throw Error('keys need to be formatted in [string] or non-empty string');
};

export default {
	removeObjectKeys,
};
