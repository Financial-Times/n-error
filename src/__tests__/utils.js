import { removeObjectKeys } from '../utils';

describe('removeObjectKeys', () => {
	describe('remove keys from object', () => {
		it('with [string] input', () => {
			const obj = { a: 1, b: 2, c: 'test', 'more-complex': 'test' };
			const removeKeyList = ['a', 'more-complex'];
			const result = removeObjectKeys(obj)(removeKeyList);
			expect(result).toEqual({ b: 2, c: 'test' });
		});

		it('non-empty string input', () => {
			const obj = { a: 1, b: 2, c: 'test', 'more-complex': 'test' };
			const result = removeObjectKeys(obj)('more-complex');
			expect(result).toEqual({ a: 1, b: 2, c: 'test' });
		});
	});

	describe('should work correctly for', () => {
		it('empty input array', () => {
			const obj = { a: 1, b: 2, c: 'test', 'more-complex': 'test' };
			const removeKeyList = [];
			const result = removeObjectKeys(obj)(removeKeyList);
			expect(result).toEqual(obj);
		});

		it('array construction with emtpy array rest spread ', () => {
			const obj = { a: 1, b: 2, c: 'test', 'more-complex': 'test' };
			const removeKeyList = [...[], 'test', 'more-complex'];
			const result = removeObjectKeys(obj)(removeKeyList);
			expect(result).toEqual({ a: 1, b: 2, c: 'test' });
		});

		it('input array with empty string', () => {
			const obj = { a: 1, b: 2, c: 'test', 'more-complex': 'test' };
			const removeKeyList = [''];
			const result = removeObjectKeys(obj)(removeKeyList);
			expect(result).toEqual(obj);
		});
	});

	describe('mutates the original input object', () => {
		class Test {
			constructor(fields) {
				Object.keys(fields).forEach(key => {
					this[key] = fields[key];
				});
			}
		}

		it('with [string] input', () => {
			const e = new Test({ status: 400, type: 'SOME_TYPE' });
			const result = removeObjectKeys(e)(['status']);
			expect(result).toMatchSnapshot();
		});

		it('non-empty string input', () => {
			const e = new Test({ status: 400, type: 'SOME_TYPE' });
			const result = removeObjectKeys(e)('status');
			expect(result).toMatchSnapshot();
		});
	});

	// describe('undefined reserved fields in Error or extended Error', () => {
	// 	it('with [string] input', () => {
	// 		const e = new Error('t');
	// 		e.type = 'SOME_TYPE';
	// 		const result = removeObjectKeys(e)(['message']);
	// 		expect(e.message).toBe('t');
	// 		expect(result.message).toBeUndefined();
	// 		expect(result.type).toBe('SOME_TYPE');
	// 	});

	// 	it('non-empty string input', () => {
	// 		const e = new Error('t');
	// 		e.type = 'SOME_TYPE';
	// 		const result = removeObjectKeys(e)('message');
	// 		expect(e.message).toBe('t');
	// 		expect(result.message).toBeUndefined();
	// 		expect(result.type).toBe('SOME_TYPE');
	// 	});
	// });

	it('throw error if keys input is not an array or string', () => {
		const obj = { a: 1, b: 2, c: 'test', 'more-complex': 'test' };
		const removeKeyList = '';
		const wrongInput = () => removeObjectKeys(obj)(removeKeyList);
		expect(wrongInput).toThrowErrorMatchingSnapshot();
	});
});
