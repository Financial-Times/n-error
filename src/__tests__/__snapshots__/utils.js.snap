// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`removeObjectKeys mutates the original input object non-empty string input 1`] = `
Test {
  "type": "SOME_TYPE",
}
`;

exports[`removeObjectKeys mutates the original input object with [string] input 1`] = `
Test {
  "type": "SOME_TYPE",
}
`;

exports[`removeObjectKeys throw error if keys input is not an array or string 1`] = `"keys need to be formatted in [string] or non-empty string"`;
